clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -686,22 +686,47 @@ static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
                                  CXXRecordDecl *Wrapper,
                                  Handlers &... handlers);
 
+template <typename RangeTy, typename... Handlers>
+static void VisitField(CXXRecordDecl *Owner, RangeTy Item, QualType ItemTy,
+                       Handlers &... handlers) {
+  if (Util::isSyclAccessorType(ItemTy)) {
+    (void)std::initializer_list<int>{
+        (handlers.handleSyclAccessorType(Item, ItemTy), 0)...};
+  } else if (Util::isSyclStreamType(ItemTy))
+    (void)std::initializer_list<int>{
+        (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+  else {
+    if (ItemTy->isArrayType()) {
+      VisitArrayElements(Item, ItemTy, handlers...);
+    } else if (ItemTy->isStructureOrClassType()) {
+      VisitAccessorWrapper(Owner, Item, ItemTy->getAsCXXRecordDecl(),
+                           handlers...);
+    }
+  }
+}
+
+template <typename RangeTy, typename... Handlers>
+static void VisitArrayElements(RangeTy Item, QualType FieldTy,
+                               Handlers &... handlers) {
+  const ConstantArrayType *CAT = cast<ConstantArrayType>(FieldTy);
```

Comment:
When you do `cast` to `ConstantArrayType` you expect that `FieldTy` is `ConstantArrayType`, otherwise assertion will be triggered, so, is it safe to do `cast` here? I'm asking because `VisitArrayElements` is called after check on `isArrayType`, not on `isConstantArrayType`.

Reply:
Later, he leaves this comment, which made me not comment on this. but it is still worth evaluating:
    // C++ lambda capture already flags non-constant array types.
    // Here, we check copyability.

There are 4 types of arrays:
Constant
This one is obvious.

DependentSized
I dont think we have to worry about this one, all instantiations are done before we get here, right?

Incomplete
You cannot actually capture an incomplete sized array type as far as I can tell.  They aren't particularly easy to spell in C, and I think we prevent capturing them: https://godbolt.org/z/-oDsdz

Variable
This is the only one that I think is possible, but only as a reference type: https://godbolt.org/z/hEdiK-
Of course, VLAs in structs aren't supported in clang (with the pretty assertive message: fields must have a constant size: 'variable length array in structure' extension will never be supported)

That said, Do we prevent the reference case from making it here?

Reply:
AFAIK SYCL spec allows only "by value" ([=] this one) capture for kernels, AFAIK we just don't diagnose if someone tried to use "by reference" capture. 
This comment didn't satisfy me because kernel object can be defined not only as lambda but as user-defined callable object as well.

Reply:
Well, then the VariableArraytype isn't a concern, though we should probably diagnose that (should we diagnose ALL ReferenceType fields?).

I'd not thought about other object types.  From that perspective, only the incomplete type is of concern, we should likely diagnose that.

Reply:
FYI: This is another great use of the 'have visit methods return the boolean value'.  We could/should stop descending once we report failure.

Reply:
Oh, I just forgot, but it seems we already diagnose ReferenceType fields, don't we?
https://github.com/intel/llvm/blob/08f8656280738c9e59b623ea51764d07fd844f9a/clang/lib/Sema/SemaSYCL.cpp#L810

> From that perspective, only the incomplete type is of concern, we should likely diagnose that.
Agree.

Reply:
Non-constant arrays could slip through via function objects. I've added a check.